source/lac/trilinos_tpetra_sparsity_pattern.cc

```diff
@@ -0,0 +1,1078 @@
+              // diagonal
+              if (column_space_map->getGlobalNumElements() ==
+                  graph->getRangeMap()->getGlobalNumElements())
+                column = row;
+              // if not, take a column index that we have ourselves since we
+              // know for sure it is there (and it will not create spurious
+              // messages to many ranks like putting index 0 on many processors)
+              else if (column_space_map->getLocalNumElements() > 0)
+                column = column_space_map->getGlobalElement(0);
+              nonlocal_graph->insertGlobalIndices(row, 1, &column);
+            }
+          Assert(nonlocal_graph->getRowMap()->getLocalNumElements() == 0 ||
+                   column_space_map->getGlobalNumElements() == 0,
+                 ExcInternalError());
+
+          nonlocal_graph->fillComplete(column_space_map, graph->getRangeMap());
+          Tpetra::Export<int, dealii::types::signed_global_dof_index, Node>
+            exporter(nonlocal_graph->getRowMap(), graph->getRowMap());
+          graph->fillComplete(column_space_map, graph->getRangeMap());
+        }
+      else
+        {
+          graph->globalAssemble();
+        }
+
+      // Check consistency between the sizes set at the beginning and what
+      // Trilinos stores:
+      using namespace deal_II_exceptions::internals;
+      Assert(compare_for_equality(n_rows(), graph->getGlobalNumEntries()),
+             ExcInternalError());
+      Assert(compare_for_equality(n_cols(), graph->getGlobalNumEntries()),
+             ExcInternalError());
+    }
+
+
+
+    template <typename Node>
+    bool
+    SparsityPattern<Node>::row_is_stored_locally(const size_type i) const
+    {
+      return graph->getRowMap()->getLocalElement(i) !=
+             Teuchos::OrdinalTraits<int>::invalid();
+    }
+
+
+
+    template <typename Node>
+    bool
+    SparsityPattern<Node>::exists(const size_type i, const size_type j) const
+    {
+      if (!row_is_stored_locally(i))
+        return false;
+
+      // Extract local indices in  the matrix.
+      auto trilinos_i = graph->getRowMap()->getLocalElement(i);
+      auto trilinos_j = graph->getColMap()->getLocalElement(j);
+
+      typename GraphType::local_inds_host_view_type col_indices;
+
+      // Generate the view.
+      graph->getLocalRowView(trilinos_i, col_indices);
+
+      // Search the index
+      const size_type local_col_index =
+        std::find(col_indices.data(),
+                  col_indices.data() + col_indices.size(),
+                  trilinos_j) -
+        col_indices.data();
+
+      return static_cast<size_t>(local_col_index) != col_indices.size();
+    }
+
+
+
+    template <typename Node>
+    typename SparsityPattern<Node>::size_type
+    SparsityPattern<Node>::bandwidth() const
+    {
+      size_type                         local_b  = 0;
+      TrilinosWrappers::types::int_type global_b = 0;
```

Comment:
Let's move the declaration of this variable down to its first use.

See also #16409.

Reply:
Done